docs/async-await.md

```diff
@@ -0,0 +1,168 @@
+
+We would like to explore how we could offer APIs that make use of these new language features to allow users to run HTTPRequest using these new idioms.
+
+This proposal describes what these APIs could look like and explores some of the potential usability concerns.
+
+## Proposed API additions
+
+### New `AsyncRequest` type
+
+The proposed new `AsyncRequest` shall be a simple swift structure.
+
+```swift
+struct AsyncRequest {
+  /// The requests url.
+  var url: String
+  
+  /// The request's HTTPMethod
+  var method: HTTPMethod
+  
+  /// The request's headers
+  var headers: HTTPHeaders
+
+  /// The request's body
+  var body: Body?
+    
+  init(url: String) {
+    self.url = url
+    self.method = .GET
+    self.headers = .init()
+    self.body = .none
+  }
+}
+```
+
+A notable change from the current [`HTTPRequest`][HTTPRequest] is that the url is not of type `URL`. This makes the creation of a request non throwing. Existing issues regarding current API: 
+
+- [HTTPClient.Request.url is a let constant][issue-395]
+- [refactor to make request non-throwing](https://github.com/swift-server/async-http-client/pull/56)
+- [improve request validation](https://github.com/swift-server/async-http-client/pull/67)
+
+The url validation will become part of the normal request validation that occurs when the request is scheduled on the `HTTPClient`. If the user supplies a request with an invalid url, the http client, will reject the request.
+
+In normal try/catch flows this should not change the control flow:
+
+```swift
+do {
+  var request = AsyncRequest(url: "invalidurl")
+  try await httpClient.execute(request, deadline: .now() + .seconds(3))
+} catch {
+  print(error)
+}
+```
+
+If the library code throws from the AsyncRequest creation or the request invocation the user will, in normal use cases, handle the error in the same catch block. 
+
+#### Body streaming
+
+The new AsyncRequest has a new body type, that is wrapper around an internal enum. This allows us to evolve this type for use-cases that we are not aware of today. 
+
+```swift
+public struct Body {
+  static func bytes<S: Sequence>(_ sequence: S) -> Body where S.Element == UInt8
+
+  static func stream<S: AsyncSequence>(_ sequence: S) -> Body where S.Element == ByteBuffer
+
+  static func stream<S: AsyncSequence>(_ sequence: S) -> Body where S.Element == UInt8
+}
+```
+
+The main difference to today's `Request.Body` type is the lack of a `StreamWriter` for streaming scenarios. The existing StreamWriter offered the user an API to write into (thus the user was in control of when writing happened). The new `AsyncRequest.Body` uses `AsyncSequence`s to stream requests. By iterating over the provided AsyncSequence, the HTTPClient is  in control when writes happen, and can ask for more data efficiently. 
+
+Using the `AsyncSequence` from the Swift standard library as our upload stream mechanism dramatically reduces the learning curve for new users.
```

Comment:
I propose to reword the above about "losing" the writer approach, since we still have it -- thanks to AsyncStream, this way this is only simplification without feature loss 👍